pkg/dynamicinstrumentation/diconfig/location_expression.go

```diff
@@ -0,0 +1,322 @@
+							ditypes.CopyLocationExpression(),
+							ditypes.ApplyOffsetLocationExpression(uint(int(elementParam.ParameterPieces[0].TotalSize)*i)),
+						)
+						targetExpressions = append(targetExpressions, expressionsToUseForEachArrayElement...)
+					}
+				} else {
+					targetExpressions = append(targetExpressions,
+						ditypes.DereferenceToOutputLocationExpression(uint(elementParam.TotalSize)))
+				}
+			}
+		}
+		expressionTargets[i].Parameter.LocationExpressions = targetExpressions
+		targetExpressions = []ditypes.LocationExpression{}
+	}
+}
+
+func collectAllLocationExpressions(parameter *ditypes.Parameter, remove bool) []ditypes.LocationExpression {
+	if parameter == nil {
+		return []ditypes.LocationExpression{}
+	}
+	expressions := parameter.LocationExpressions
+	for i := range parameter.ParameterPieces {
+		expressions = append(expressions, collectAllLocationExpressions(parameter.ParameterPieces[i], remove)...)
+	}
+	if remove {
+		parameter.LocationExpressions = []ditypes.LocationExpression{}
+	}
+	return expressions
+}
+
+type ExpressionParamTuple struct {
+	TypePath  string
+	Parameter *ditypes.Parameter
+}
+
+// generateLocationVisitsMap follows the tree of parameters (parameter.ParameterPieces), and
+// collects string values of all the paths to nodes that need expressions (`needsExpressions`),
+// as well as all combinations of elements that can be achieved by walking the tree (`trieKeys`).
+func generateLocationVisitsMap(parameter *ditypes.Parameter) (trieKeys, needsExpressions []ExpressionParamTuple) {
+	trieKeys = []ExpressionParamTuple{}
+	needsExpressions = []ExpressionParamTuple{}
+
+	var visit func(param *ditypes.Parameter, path string)
+	visit = func(param *ditypes.Parameter, path string) {
+		if param == nil {
+			return
+		}
+		trieKeys = append(trieKeys, ExpressionParamTuple{path + param.Type, param})
+
+		if (len(param.ParameterPieces) == 0 ||
+			isBasicType(param.Kind) ||
+			param.Kind == uint(reflect.Array) ||
+			param.Kind == uint(reflect.Slice)) &&
+			param.Kind != uint(reflect.Struct) &&
+			param.Kind != uint(reflect.Pointer) {
+			needsExpressions = append(needsExpressions, ExpressionParamTuple{path + param.Type, param})
+			return
+		}
+
+		for i := range param.ParameterPieces {
+			newPath := path + param.Type + "@"
+			visit(param.ParameterPieces[i], newPath)
+		}
+	}
+	visit(parameter, "")
+	return trieKeys, needsExpressions
+}
+
+func isBasicType(kind uint) bool {
+	switch reflect.Kind(kind) {
+	case reflect.Bool, reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
+		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr,
+		reflect.Float32, reflect.Float64, reflect.Complex64, reflect.Complex128, reflect.String:
+		return true
+	default:
+		return false
+	}
+}
+
+func randomLabel() string {
+	length := 6
+	randomString := make([]byte, length)
+	for i := 0; i < length; i++ {
+		randomString[i] = byte(65 + rand.Intn(25))
```

Comment:
What is `65` here? If you are just needing a random alphanumeric label, then this code works well:
```
func randomLabel() string {
    length := 6
	buf := make([]byte, length/2)
	_, _ = rand.Read(buf)
	return hex.EncodeToString(buf)
}
```